File: backend-on-python/app/modules/tests/tests_controller.py
from typing import List

from fastapi import APIRouter, Depends, HTTPException

from app.core.models.users_model import User
from app.core.utils.auth_utils import get_current_user
from app.modules.tests.tests_schemas import (
    TestSummaryResponse,
    TestDetailsResponse,
    TestQuestionResponse,
    TestQuestionSubmitRequest,
    PersonalityAnalysisResponse,
)
from app.modules.tests.tests_service import TestsService

router = APIRouter()


@router.get("/", response_model=List[TestSummaryResponse])
def get_tests(
    current_user: User = Depends(get_current_user),
    service: TestsService = Depends(),
):
    return service.get_all_tests_for_user(current_user.id)


@router.get("/{test_id}", response_model=TestDetailsResponse)
def get_test_details(
    test_id: int,
    current_user: User = Depends(get_current_user),
    service: TestsService = Depends(),
):
    return service.get_test_details(current_user.id, test_id)


@router.get("/{test_id}/questions/{question_id}", response_model=TestQuestionResponse)
def get_test_question(
    test_id: int,
    question_id: int,
    service: TestsService = Depends(),
):
    return service.get_test_question(test_id, question_id)


@router.post("/{test_id}/questions/{question_id}")
def submit_answer(
    test_id: int,
    question_id: int,
    data: TestQuestionSubmitRequest,
    current_user: User = Depends(get_current_user),
    service: TestsService = Depends(),
):
    service.submit_answer(current_user.id, test_id, question_id, data)
    return {"message": "Answer submitted successfully"}


@router.post("/analysis", response_model=PersonalityAnalysisResponse)
def analyze_tests(
    current_user: User = Depends(get_current_user),
    service: TestsService = Depends(),
):
    return service.analyze_tests(current_user.id)

File: src/modules/tests/tests.controller.ts
import {
	Controller,
	Get,
	Post,
	Param,
	Body,
	UseGuards,
	ParseIntPipe,
	HttpCode,
	HttpStatus,
	NotFoundException,
} from '@nestjs/common'
import { ApiTags } from '@nestjs/swagger'
import { GetCurrentUser } from 'src/common/decorators/get-current-user.decorator'
import { UserAuthGuard } from 'src/common/guards/user-auth.guard'
import { User } from '@prisma/client'
import { TestsService } from './tests.service'
import {
	TestSummaryDto,
	TestDetailsDto,
	TestQuestionDto,
	TestQuestionSubmitDto,
	PersonalityAnalysisDto,
} from './dtos/tests.dto'

@ApiTags('tests')
@Controller('tests')
export class TestsController {
	constructor(private readonly testsService: TestsService) {}

	@Get('/')
	@UseGuards(UserAuthGuard)
	@HttpCode(HttpStatus.OK)
	async getAllTestsForUser(@GetCurrentUser() user: User): Promise<TestSummaryDto[]> {
		return this.testsService.getAllTestsForUser(user.id)
	}

	@Get(':testId')
	@UseGuards(UserAuthGuard)
	@HttpCode(HttpStatus.OK)
	async getTestDetails(
		@Param('testId', ParseIntPipe) testId: number,
		@GetCurrentUser() user: User,
	): Promise<TestDetailsDto> {
		const testDetails = await this.testsService.getTestDetails(user.id, testId)
		if (!testDetails) {
			throw new NotFoundException('Test not found')
		}
		return testDetails
	}

	@Get(':testId/questions/:questionId')
	@HttpCode(HttpStatus.OK)
	async getTestQuestion(
		@Param('testId', ParseIntPipe) testId: number,
		@Param('questionId', ParseIntPipe) questionId: number,
	): Promise<TestQuestionDto> {
		const question = await this.testsService.getTestQuestion(testId, questionId)
		if (!question) {
			throw new NotFoundException('Question not found')
		}
		return question
	}

	@Post(':testId/questions/:questionId')
	@UseGuards(UserAuthGuard)
	@HttpCode(HttpStatus.OK)
	async submitAnswer(
		@Param('testId', ParseIntPipe) testId: number,
		@Param('questionId', ParseIntPipe) questionId: number,
		@Body() data: TestQuestionSubmitDto,
		@GetCurrentUser() user: User,
	): Promise<{ message: string }> {
		return this.testsService.submitAnswer(user.id, testId, questionId, data)
	}

	@Post('analysis')
	@UseGuards(UserAuthGuard)
	@HttpCode(HttpStatus.OK)
	async analyzeTests(@GetCurrentUser() user: User): Promise<PersonalityAnalysisDto> {
		return this.testsService.analyzeTests(user.id)
	}
}
